src/common/transformations/src/transformations/common_optimizations/convert_u4_weights_zero_point_to_scalar.cpp

```diff
@@ -0,0 +1,80 @@
+#include "transformations/utils/utils.hpp"
+
+ov::pass::ConvertU4WeightsZeroPointToScalar::ConvertU4WeightsZeroPointToScalar() {
+    MATCHER_SCOPE(ConvertU4WeightsZeroPointToScalar);
+    auto weights_m = pattern::wrap_type<ov::op::v0::Constant>(pattern::type_matches(ov::element::u4));
+    auto convert_m = pattern::wrap_type<ov::op::v0::Convert>({weights_m}, pattern::consumers_count(1));
+
+    auto float_zp_predicate = [](ov::Output<ov::Node> output) -> bool {
+        return pattern::type_matches_any({ov::element::f32, ov::element::f16})(output) &&
+               pattern::consumers_count(1)(output);
+    };
+    auto float_zero_point_m = pattern::wrap_type<ov::op::v0::Constant>(float_zp_predicate);
+
+    auto u4_zp_predicate = [](ov::Output<ov::Node> output) -> bool {
+        return pattern::type_matches(ov::element::u4)(output) && pattern::consumers_count(1)(output);
+    };
+    auto u4_zero_point_m = pattern::wrap_type<ov::op::v0::Constant>(u4_zp_predicate);
+    auto zero_point_convert_m = pattern::wrap_type<ov::op::v0::Convert>({u4_zero_point_m}, float_zp_predicate);
+
+    auto zero_point_m = std::make_shared<pattern::op::Or>(OutputVector{float_zero_point_m, zero_point_convert_m});
+    auto subtract_m = pattern::wrap_type<ov::op::v1::Subtract>({convert_m, zero_point_m});
+
+    ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        auto weights = ov::as_type_ptr<ov::op::v0::Constant>(pattern_map.at(weights_m).get_node_shared_ptr());
+        std::shared_ptr<ov::op::v0::Constant> zero_point;
+        if (pattern_map.count(float_zero_point_m)) {
+            const auto& float_zp = pattern_map.at(float_zero_point_m);
+            zero_point = ov::as_type_ptr<ov::op::v0::Constant>(float_zp.get_node_shared_ptr());
+        } else {
+            const auto& u4_zp = pattern_map.at(u4_zero_point_m);
+            zero_point = ov::as_type_ptr<ov::op::v0::Constant>(u4_zp.get_node_shared_ptr());
+        }
+        if (!weights || !zero_point)
+            return false;
+        // Due to the matcher specific and Subtract branches similarity,
+        // weights and zero_point might be mixed up with each other
+        if (ov::shape_size(weights->get_shape()) < ov::shape_size(zero_point->get_shape()))
+            std::swap(zero_point, weights);
+
+        auto zero_point_shape = zero_point->get_shape();
+        if (ov::shape_size(zero_point_shape) == 1)
+            return false;
+
+        const auto& weights_shape = weights->get_shape();
+        const size_t weights_rank = weights_shape.size();
+        const size_t zero_point_rank = zero_point_shape.size();
+        // Zero point constant can be converted into scalar only if this does not affect Subtract output shape
+        if (weights_rank < zero_point_rank)
+            return false;
+
+        zero_point_shape.insert(zero_point_shape.begin(), weights_rank - zero_point_rank, 1);
```

Comment:
It seems "zero_point_shape.insert" is not necessary, since shape comparison can be done without insertion and zero_point_shape won't be needed after that.

Reply:
@xuchen-intel the comment is applied in #20709. Could you please review the PR? Thanks